osu.Game/Database/RealmContextFactory.cs

```diff
@@ -0,0 +1,160 @@
+
+        /// <summary>
+        /// Lock object which is held for the duration of a write operation (via <see cref="GetForWrite"/>).
+        /// </summary>
+        private readonly object writeLock = new object();
+
+        private ThreadLocal<bool> refreshCompleted = new ThreadLocal<bool>();
+
+        private bool rollbackRequired;
+
+        private int currentWriteUsages;
+
+        private Transaction currentWriteTransaction;
+
+        public RealmContextFactory(Storage storage)
+        {
+            this.storage = storage;
+
+            threadContexts = new ThreadLocal<Realm>(createContext, true);
+
+            // creating a context will ensure our schema is up-to-date and migrated.
+            var realm = Get();
+            Logger.Log($"Opened realm \"{realm.Config.DatabasePath}\" at version {realm.Config.SchemaVersion}");
+        }
+
+        private void onMigration(Migration migration, ulong lastSchemaVersion)
+        {
+        }
+
+        private static readonly GlobalStatistic<int> reads = GlobalStatistics.Get<int>("Realm", "Get (Read)");
+        private static readonly GlobalStatistic<int> writes = GlobalStatistics.Get<int>("Realm", "Get (Write)");
+        private static readonly GlobalStatistic<int> commits = GlobalStatistics.Get<int>("Realm", "Commits");
+        private static readonly GlobalStatistic<int> rollbacks = GlobalStatistics.Get<int>("Realm", "Rollbacks");
+        private static readonly GlobalStatistic<int> contexts_open = GlobalStatistics.Get<int>("Realm", "Contexts (Open)");
+        private static readonly GlobalStatistic<int> contexts_created = GlobalStatistics.Get<int>("Realm", "Contexts (Created)");
+
+        /// <summary>
+        /// Get a context for the current thread for read-only usage.
+        /// If a <see cref="RealmWriteUsage"/> is in progress, the existing write-safe context will be returned.
+        /// </summary>
+        public Realm Get()
+        {
+            reads.Value++;
+            return getContextForCurrentThread();
+        }
+
+        /// <summary>
+        /// Request a context for write usage. Can be consumed in a nested fashion (and will return the same underlying context).
+        /// This method may block if a write is already active on a different thread.
+        /// </summary>
+        /// <returns>A usage containing a usable context.</returns>
+        public RealmWriteUsage GetForWrite()
+        {
+            writes.Value++;
+            Monitor.Enter(writeLock);
+            Realm context;
+
+            try
+            {
+                context = getContextForCurrentThread();
+
+                currentWriteTransaction ??= context.BeginWrite();
+            }
+            catch
+            {
+                // retrieval of a context could trigger a fatal error.
+                Monitor.Exit(writeLock);
+                throw;
+            }
+
+            Interlocked.Increment(ref currentWriteUsages);
```

Comment:
isn't this a little overkill? the monitor is taken at this point, so there surely can be no concurrency here?

Reply:
No idea, it's what we had in `DatabaseContextFactory`. Again, not sure if we need any of this, rather than relying on realm's transaction directly (the reason we made this was to allow nested transactions though, for cases like `BeatmapManager` -> `FileStore`).

Reply:
was not aware that this duplicates/mirrors the existing db class, fair enough